v1.1 -> v2.0 Semantics refactor

ejhumphrey opened the pull request:
Renames tests for clearer mappings between submodules
Addresses much of the conversation from #75 
- will break the current API, and needs some help signaling deprecations
- `Streamer.generate` is now `Streamer.iterate`, since it returns an iterator
- `Streamer`types are now directly iterable
- `Mux` semantics are simplified to remove the (now superfluous) `seed` concept, which were synonymous with streams; now there are only streamers and streams.
- the notion of `batch` is scrubbed in all core functionality.

Will resolve: #64, #75, #76, #77, #78, #79, #81

bmcfee commented on pescador/buffered.py:
```diff
@@ -56,31 +58,31 @@ def activate(self):
         self.stream_ = self.streamer
 
-    def generate(self, max_batches=None):
+    def iterate(self, max_iter=None):
```
persist the old generate function here.  we can wrap the call signature for API compatibility, but there should be a warning.

ejhumphrey replied:
roger that

ejhumphrey replied:
@bmcfee do you have an example of "@move"ing a class's method? I could hack it out of if need be, but I'd rather do the fast thing.

bmcfee commented on pescador/buffered.py:
```diff
@@ -56,31 +58,31 @@ def activate(self):
-def buffer_batch(generator, buffer_size):
-    '''Buffer an iterable of batches into larger (or smaller) batches
+def buffer_batch(streamer, buffer_size):
+    '''Buffer data samples from an streamer into one data object.
```
this doesn't actually require a streamer, right?

ejhumphrey replied:
no, it needs an iterable; not sure what a sane variable name would be here. I also aspire to change this on a subsequent PR, so .... any strong opinions?

ejhumphrey replied:
I've opted for `stream`

bmcfee commented on pescador/core.py:
```diff
@@ -8,6 +8,11 @@
 
 class StreamActivator(object):
     def __init__(self, streamer):
+        for mname in ['activate', 'deactivate']:
```
why not just check `isinstance(streamer, Streamer)` instead?

ejhumphrey replied:
**tl;dr:** ya, that's better.

**commentary:** I think I had a lapse of Java ptsd, and wanted to check that the object implements the interface of `activate`/`deactivate`, since it doesn't strictly matter that it's a Streamer (although when would we run into that? /shrug). If we really cared, a more Pythonic way of doing this would be to define an ActiveStream mixin, but ... this is total overkill for where we are right now, and mention it now only for completeness.

bmcfee commented on pescador/core.py:
```diff
@@ -92,8 +98,8 @@ def __init__(self, streamer, *args, **kwargs):
-            raise PescadorError('streamer must be a generator, iterable, or '
-                                'Streamer')
+            raise PescadorError('`streamer` must be a callable function that '
+                                'returns an iterable object.')
```
could we actually generalize this and support iterables as well as functions?  Call it a stretch goal.

ejhumphrey replied:
Actually it does as of [this change](https://github.com/pescadores/pescador/pull/80/files/44d82c225dd8ceb6777cead497a7682ee4eb195c#diff-535170d9182a6f744ecc1f98d69624d5R124) 😄  ... I've updated the error message, and added a `Streamer(iterable)` test in da42dbf1.

bmcfee commented on pescador/core.py:
```diff
@@ -113,57 +119,53 @@ def activate(self):
 
-    def generate(self, max_batches=None):
-        '''Instantiate the generator
+    def iterate(self, max_iter=None):
```
again, deprecation rename / persist old signature with warning

ejhumphrey replied:
roger roger

bmcfee commented on pescador/core.py:
```diff
@@ -113,57 +119,53 @@ def activate(self):
+    #       into a "transform". @bmcfee please advise? example to follow?
     def tuples(self, *items, **kwargs):
         '''Generate data in tuple-form instead of dicts.
 
```
I'm still not 100% on what a `transform` is defined to be.

I find the `.tuples()` interface quite useful for interfacing with keras though, and I'd prefer to not have to replace it with something more cumbersome.

ejhumphrey replied:
I've added a [long-form response here](https://github.com/pescadores/pescador/issues/77#issuecomment-292806012), including why I think this is an anti-pattern.

**tl;dr:** `Streamer.tuples` only works for a subset of data streams that might be fed through it; I'd propose explicitly separating `Streamer` functionality from processing data streams. 

The thing I'd like to replace it with is a separate `tuples` function (as sketched out in that linked issue), which I'd argue isn't any more cumbersome than what we currently have and Streamer remains data stream agnostic.

bmcfee commented on pescador/mux.py:
```diff
@@ -19,24 +19,26 @@ class Mux(core.Streamer):
-                 lam=256.0, pool_weights=None, with_replacement=True,
-                 prune_empty_seeds=True, revive=False,
+    def __init__(self, streamers, k,
+                 lam=256.0, weights=None, with_replacement=True,
```
deprecation renames EVERYWHERE

while we're at it though, I'd like to replace `lam`=>`rate`

ejhumphrey replied:
haha roger

bmcfee commented on pescador/zmq_stream.py:
```diff
@@ -81,15 +81,16 @@ def zmq_recv_batch(socket, flags=0, copy=True, track=False):
 
 
-def zmq_worker(port, streamer, terminate, copy=False, max_batches=None):
+def zmq_worker(port, streamer, terminate, copy=False, max_iter=None):
```
renames

ejhumphrey replied:
yayaya

bmcfee commented:
@ejhumphrey Re: handling variable rename deprecations, here's how librosa does it.  Note that the current stable version has no lingering renames, but rolling back to 0.4.3 you can find some examples.

1. Check out the `util.deprecation` [submodule](https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/util/deprecation.py)
2. Here's an example of it in use: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/core/constantq.py#L327-L329
3. There's also some helpers in the `util.decorators` [submodule](https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/util/decorators.py) for deprecating entire functions, or moving them to different submodules/renaming them.
4. Here's an example of move/rename behavior: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/core/audio.py#L692-L693
5. Here's an example of deprecated function behavior: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/feature/spectral.py#L1337-L1338

The [futurepast](https://github.com/amueller/futurepast) package implements some of these things and more, but it's not yet at an initial release.

ejhumphrey commented:
thanks @bmcfee, made it pretty far ... still have some work to do, but any advice / pointers will be helpful.

bmcfee commented:
@ejhumphrey I don't have an example of `@move`ing a class method, because that gets a bit more involved. However, I think you don't necessarily want or need that here, since the call signature has changed as well.  I would just write a shim function to translate to the new API, and a custom warning message.

cjacoby commented:
(I have made a first pass looking through this; I won't have time for another look until tomorrow, because work+taxes)

ejhumphrey commented:
guh, I need to get the test coverage back up (idk what's up with that), but this is in a state where it's worth taking another look @bmcfee ... I think I did most of the deprecation stuff correctly, but do let me know if that's not the case (I have no doubt you will :o)

bmcfee commented on pescador/maps.py:
```diff
@@ -0,0 +1,120 @@
+        inputs = [inputs]
+    if outputs and isinstance(outputs, six.string_types):
+        outputs = [outputs]
+
```
I don't think we should upcast to list here.  Keras's interface is overloaded to support tensors or lists of tensors as inputs (or outputs). The following are all perfectly valid:

```python
model = K.models.Model(x1, y1)
model = K.models.Model(x1, [y1, y2])
model = K.models.Model([x1, x2], [y1, y2])
```
where `x*` and `y*` are tensors.

`keras_tuples` should let me do any of the following:

```python
keras_tuples(streamer, inputs='X1', outputs='Y1')  # generates tuples of (ndarray, ndarray)
keras_tuples(streamer, inputs='X1', outputs=['Y1', 'Y2'])  # generates tuples of (ndarray, list of ndarray)
keras_tuples(streamer, inputs=['X1'], outputs=['Y1', 'Y2'])  # generates tuples of (list of ndarray, list of ndarray)
```

ejhumphrey replied:
cool, this is helpful. I'm not super familiar with the `fit` interface, so I was trying to get as far as possible from the docs.

bmcfee replied:
ahahhahahahhaha

![Documentation?](https://media.giphy.com/media/YWZvYDKmnkZNu/giphy.gif)

ejhumphrey replied:
now that I'm reading this on a computer, rather than my phone, I've discovered that I've written the behavior you desire ... although perhaps in not the most elegant way. 

The *keys* get upcast to a list, but singleton items are downcast to np.ndarrays; see L112, L116.

This was the first semi-clean way I saw toward the end of the night to handle both string / list inputs, and return np.ndarray / list outputs.

bmcfee replied:
I think we also want an old-style tuples map that works like `Streamer.tuples` does/did:

```python
def tuples(gen, *args):
    for item in gen:
         yield (item[x] for x in args)
```
or something

bmcfee replied:
Upon closer inspection, I think this still isn't correct.

I don't think `fit(inputs=x)` and `fit(inputs=[x])` do the same thing (I might be wrong on that though).  If that's the case, then upcast-downcast will not behave as expected.

bmcfee commented on pescador/zmq_stream.py:
```diff
@@ -181,18 +180,19 @@ def generate(self, max_batches=None):
         except StopIteration:
             pass
 
+        # TODO: Fix bare except
```
no, this is an intentional bare except that gets reraised.  IIRC, this is here because of some weirdness with transportable exceptions.

ejhumphrey replied:
gotcha, I'll nuke the comment ... I was just following my linter blind

bmcfee replied:
feel free to add a pylint disable comment there.

ejhumphrey commented:
Alrighty, I think this PR is all buttoned up; tests pass, and fixed the drop in coverage. Just need signoff from y'all (@bmcfee & @cjacoby)

cjacoby commented:
Apologies; I'm running around like crazy this week in preparation for flying to NYC tomorrow. If i don't get to it today (unlikely, honestly), I'll catch up on all this on the plane tomorrow. 

If that's blocking you and you want to push ahead, either just do it or poke me "offline" and I'll bump it up in my queue.

bmcfee commented on pescador/buffered.py:
```diff
@@ -46,6 +48,12 @@ def __init__(self, streamer, buffer_size,
+             '`pescador.buffer_data` in 2.0.0, which can be used with '
+             '`pescador.Streamer` to similar ends.'
+             'Use this idiom instead to maintain forwards compatibility.',
+             DeprecationWarning)
```
Are the examples in the documentation up to date with the latest jazz?

ejhumphrey replied:
no, probably not. I'll look into that now

bmcfee commented on pescador/core.py:
```diff
@@ -35,54 +41,54 @@ class Streamer(object):
+    >>> for i in stream(cycle=True):
+    ...     print(i)  # Displays 0, 1, 2, 3, 4, 0, 1, 2, ...
+
+    NOTE: We don't use cycle because it's a closure and can't be serialized.
```
this note shouldn't be in the docstring

ejhumphrey replied:
ha, right. that was a note for me to migrate into documentation elsewhere

bmcfee commented on pescador/core.py:
```diff
@@ -213,40 +225,49 @@ def tuples(self, *items, **kwargs):
         '''
+        max_iter = rename_kw('max_batches', max_batches,
+                             'max_iter', max_iter,
+                             '1.1.0', '2.0.0')
```
Please be consistent about major/minor version numbers in deprecations.  I see `1.1` and `2.0` elsewhere.

ejhumphrey replied:
is 1.1 fine? or 1.1.0?

bmcfee replied:
i'd go with 1.1; sub-minors should never be involved in deprecations.

bmcfee commented on pescador/core.py:
```diff
@@ -204,7 +213,10 @@ def tuples(self, *items, **kwargs):
         '''
-
+        warn('`Streamer.tuples()` is deprecated in 1.1.0 '
+             'This functionality is moved to `pescador.tuples` in 2.0.0. '
```
is there actually a `pescador.tuples`?  I only see `maps.keras_tuples`

ejhumphrey replied:
nope, that's stale

bmcfee commented on tests/test_core.py:
```diff
@@ -1,23 +1,53 @@
-def test_streamer_list():
+def test_StreamActivator():
+    with pytest.raises(pescador.core.PescadorError):
+        pescador.core.StreamActivator(range)
```
this only tests the failure mode?  I guess that's okay.

bmcfee commented on tests/test_maps.py:
```diff
@@ -0,0 +1,62 @@
+    stream = pescador.maps.keras_tuples(data, outputs="whiz")
+    for n, (x, y) in enumerate(stream):
+        assert n == y / 2.
+        assert x is None
```
Should this be` None` or `[]`?

ejhumphrey replied:
[keras docs](https://keras.io/models/model/#fit) (currently) say `None`

bmcfee commented on examples/frameworks/keras_example.py:
```diff
@@ -93,71 +117,99 @@ def build_model(input_shape):
-print('Test accuracy:', score[1])
+try:
+    model.fit_generator(
+        pescador.tuples(batches, 'X', 'y'),
```
`pescador.tuples(batches.cycle(), 'X', 'y')` ?

ejhumphrey replied:
batches is infinite anyways?

ejhumphrey commented:
alrighty, updated the examples, and added some slightly more meaningful benchmarking to the zmq_example, which burns cycles to prove that things are happening in the background.

LET'S DO THIS

bmcfee commented on examples/frameworks/keras_example.py:
```diff
@@ -93,71 +118,101 @@ def build_model(input_shape):
+    '''
     X = np.atleast_2d(X)
-    # y's are binary vectors, and should be of shape (1, 10) after this.
+    # y's are binary vectors, and should be of shape (10,) after this.
```
atleast_2d will make them shape (1,10)

ejhumphrey replied:
done

bmcfee commented on examples/mux_files_example.py:
```diff
@@ -61,14 +61,14 @@ def npz_generator(npz_path):
     npz_data = np.load(npz_path)
     X = npz_data['X']
-    # y's are binary vectors, and should be of shape (1, 10) after this.
+    # y's are binary vectors, and should be of shape (10,) after this.
```
same comment here, i think?

ejhumphrey replied:
done

cjacoby commented on pescador/mux.py:
```diff
@@ -83,14 +90,48 @@ def __init__(self, seed_pool, k,
+            Use the `streamers` parameter instead.
+            The `seed_pool` parameter will be removed in pescador 2.0.
+
+        lam : float > 0.0
```
:100: on converting `lam` to `rate` - I think that's sooo much more intuitive.

However, you missed including `rate` in the docs here. I suspect the `lam` above is supposed to be `rate`.

ejhumphrey replied:
yup, caught it

cjacoby commented on pescador/version.py:
```diff
@@ -3,4 +3,4 @@
 
 short_version = '1.0'
-version = '1.0.0'
+version = '1.1.0'
```
@bmcfee - Correct me if I misunderstand the case, but I believe last time I tried to do this that you requested all version increments be in a separate PR. I would expect that to apply here.

Also... shouldn't we increment the `short_version` too?

ejhumphrey replied:
can I trouble y'all for consensus? I'd love to get this over the goal line today 😄

cjacoby replied:
Plz see: https://github.com/pescadores/pescador/blob/master/CONTRIBUTING.md, "Additional Comments".

ejhumphrey replied:
roger roger

cjacoby commented on tests/test_buffered.py:
```diff
@@ -0,0 +1,60 @@
+
+    T.__eq_lists(reference, estimate)
+
+    estimate = pescador.BufferedStreamer(gen_stream, buf_size)
```
curious; is this copypasta, or have you kept the old interface here on purpose? Same below.

(If these tests are here to make sure the old interface still works, that's fine, just mention that in a comment someplace, and that they should be removed when the deprecations are removed.)

ejhumphrey replied:
ya, the whole `buffered` submodule is going away in 2.0, I've added a comment.

ejhumphrey commented on pescador/buffered.py:
```diff
@@ -56,151 +63,30 @@ def activate(self):
             not the number of individual samples.
+
+        partial : bool, default=True
+            Return buffers smaller than the requested size.
```
done

ejhumphrey commented:
couple comments here because this thread is getting *insane*.
- versioning: what's our stance (this vs separate)?
- docs: help would be great. also I dunno what I'm doing. so help would be *really* great.

cjacoby commented:
Re: Versioning. I'm posting this here, too 'cause embedded is getting too crazy.

The `CONTRIBUTING.md` says:
```Please keep branches as simple as possible: do not implement multiple features in the same branch, unless they are mutually dependent on one another. By the same token, please do not change unrelated portions of the repository (e.g., version numbers or changelog entries).```
and by that, I'd say, separate PR for version update.

Re: Docs - let's start an Issue with a checklist?

cjacoby commented:
(Once we have those two things, I think this PR is 👍

ejhumphrey commented:
hey team, two things: 
 - for clarity, you're saying create a docs issue **and** fix it in this PR? or create an issue and fix it elsewhere?
- in parallel, I started updating a recent project to (what will be) pescador 1.1, and having some difficulties dealing with `Mux`, specifically, the `Mux-of-Muxes` case that remains a little bit suspicious, as far as I'm concerned. I'm still in the process of determining whether my current struggles are a result of poor initial design of data streams in that project, a fundamental issue with the Mux API, or some combination of both. Regardless, I don't want to merge this PR until I've got concrete answers about what the hell is going on.

cjacoby commented:
* Re: docs. I'm proposing it gets fixed in a separate PR. The items could get logged in that PR, or in an issue. But I'd start with an issue at least. If you write up an issue with all the components that changed such that they will affect the docs, I could take a stab at updating the docs to reflect the changes made in a PR. 

* Re: Muxes. This sounds like a good candidate for identifying via improved testing, ya?

ejhumphrey commented on pescador/mux.py:
```diff
@@ -19,59 +20,65 @@ class Mux(core.Streamer):
             may not imply future emptiness.
 
         revive: bool
             If ``with_replacement`` is ``False``, setting ``revive=True``
```
@bmcfee @cjacoby, pop quiz ... if `with_replacement=True`, what is the significance of `revive`?

cjacoby replied:
Nothing.

ejhumphrey commented:
@mcfee, can I trouble you for a ✅ ?

bmcfee commented on pescador/mux.py:
```diff
@@ -104,45 +145,48 @@ def __init__(self, seed_pool, k,
 
+            # TODO: Stream uniqueness can't be guaranteed here.
+            # Other machinery necessary to impose permutation, this is
+            # implicitly `with_replacement`
```
really?  I thought we zero out the probability of active streams when `with_replacement=False`.

ejhumphrey replied:
You're correct, and I've already removed the comment locally; this was my read at midnight last night, and fixed it this am with some coffee-inspired clarity.

ejhumphrey replied:
btw I was totally wrong and just thought i did (uhderp)

ejhumphrey commented:
replaced by #88
